### ISSUE TITLE
unable to create mock provider for Jest+React testing Lib

### ISSUE THREAD

**Issue by Dujota** [bluebill1049/little-state-machine#111]
So i am trying to setup up testing for my project and all works well except when i try to create a mock/test provider for our test suite. 

*(nothing too fancy just Next.js 11 + Jest + RTL)*

The main issue is that i want to move away from mocking the return value of useStateMachine in favor of just letting the lib do its job and i test the result of the render. 

This is also hindering my ability to test click events since they rely on LSM state changes. (ie hook form controlled by lsm). 

So i have working providers for react query, i would assume that they would work the same. i did a test with just creating a React.Context  provider and that works just fine. 

So the main error i am getting is that createStore({...mockStore}) comes back with

```
// works just fine with the query provider etc. 
// When i add the provider to wrap other existing providers (ie query client for react-query) it fails

import * as React from 'react';
import { StateMachineProvider, createStore } from 'little-state-machine';
import { initialState } from '../../store/states';
import { QueryClient, QueryClientProvider } from 'react-query';

import { render } from '@testing-library/react';



const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

const mockStore = () =>
  createStore({
    ...initialState
  });

// HOC for rendering our ui components with the query client along with testing configs
export function renderWithClient(ui) {
  mockStore();
  const testQueryClient = createTestQueryClient();

  const { rerender, ...result } = render(
    <StateMachineProvider>
      <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    </StateMachineProvider>
  );
  return {
    ...result,
    rerender: (rerenderUi) =>
      rerender(
        <StateMachineProvider>
          <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>{' '}
        </StateMachineProvider>
      ),
  };
}

// Query Client Context wrapper for our tests, use this separately when we want to
// use it to render hooks outiside components for testing
export function createWrapper() {
  const testQueryClient = createTestQueryClient();
  mockStore();

  return ({ children }) => (
    <StateMachineProvider>
      <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>;
    </StateMachineProvider>
  );
}
}
```

i get the below error, which basically tells me that little state machine is undefined, and checking console confirms same. 

```
 TypeError: (0 , _littleStateMachine.createStore) is not a function

      27 |
      28 | const mockStore = () =>
    > 29 |   createStore({
         |   ^
      30 |     user,
      31 |     points,
      32 |     transactions,

      at mockStore (__mocks__/utils/index.js:29:3)
      at renderWithClient (__mocks__/utils/index.js:44:3)
```

This is how i would normally create all my mockProviders (for which all except LSM are 100% working) 

Any ideas/thoughts? Has anyone created test suites with LSM provider?

**Comment by JamesWong1999** [comment]
```tsx
describe('test', () => {
  it('should works', async () => {
    const App = () => {};
    const Root = () => {
      createStore({
        data: {}
      });

      return (
        <StateMachineProvider>
          <App />
          <App />
          <App />
          <App />
        </StateMachineProvider>
      );
    };

    render(<Root />);
  });
```

could you do something like above?

**Comment by Dujota** [comment]
> ```tsx
> describe('test', () => {
>   it('should works', async () => {
>     const App = () => {};
>     const Root = () => {
>       createStore({
>         data: {}
>       });
> 
>       return (
>         <StateMachineProvider>
>           <App />
>           <App />
>           <App />
>           <App />
>         </StateMachineProvider>
>       );
>     };
> 
>     render(<Root />);
>   });
> ```
> 
> could you do something like above?

I could try to setup a provider on each test, but the idea is to automate the boilerplate portion before each test. Works fine for all others except LSM. There's something i am missing here.

**Comment by bluebill1049** [comment]
each test probably should have its own provider, what @JamesWong1999 provided is the correct approach.

**Comment by Dujota** [comment]
> each test probably should have its own provider, what @JamesWong1999 provided is the correct approach.

So i guess what I am confused about, how does this work for all other providers but not LSM. I can make this work with React.Context since it just accepts a store. 

It doesnt even pick up the modules createStore function, so its techinically running on each test startup. (ie creates a new provider for each test case)

**Comment by bluebill1049** [comment]
That's more of a testing lib question than LSM, LSM is just leverage on top of React context API.

**Comment by Dujota** [comment]
i feel like this should work, it is pretty standard boilerplate for example:

```
function renderWithProviders(ui, { reduxState, locale = "en" }) {
  const store = createStore(reducer, reduxState || initialState);
  return (
    <ThemeProvider>
      <ReduxProvider store={store}>
        <IntlProvider locale={locale}>{ui}</IntlProvider>
      </ReduxProvider>
    </ThemeProvider>
  );
}
```

The redux example is not pretty much in line with what I was trying to do

I am going to try to refactor the _app.js providers into one `<AllProviders > {children} </AllProviders>` component and see if that can circumvent this issue

**Comment by Dujota** [comment]
I finally re-visited the topic in order to resolve this outstanding item. Originally, I had opted for a workaround by completely mocking out LSM and its return values. Although this "works" and is very usefull when you just want to programatically manipulate component ui without any simulated interactions, this still did not resolve the root problem. 

#### for example:
```
// mockLSM.js
import * as LittleStateMachine from 'little-state-machine';

  jest.mock('little-state-machine', () => ({
    useStateMachine: jest.fn(),
  }));

export const stateMachineSpy = jest.spyOn(LittleStateMachine, 'useStateMachine');

// overload the hook and allow me to return any value or action 
export function useSpyStateMachine(state = {}, actions = {}) {
  return stateMachineSpy.mockReturnValue({
    state,
    actions,
  });
}


// then in a test we use it like so:

test('should do something', () => {
    useSpyStateMachine(
      { 
        notification: {  // state.notification
          status: 'success',
          message:'it works!'
        },
      },
      {
        updateNotification: jest.fn(),  //actions.updateNotification
      }
    );
     
     // at this point the mock data is injected by jest and we can test results 

    const { container } = render(<Comp />);
    // do some test stuff + assert
  });
```

This is fine, when doing unit testing or just trying to assert some branching logic that depends on the store's current state value . Unfortunately it was not enough since I could not properly render the app in the same way a user would experience it, which was the original goal (test behaviour based on user interaction with app) . 

Again, the goal is provide a seamless, automated way to render components in the same way that the application would spin up; wrapped in LSM provider without having to add repeetitive boilerplate code to every test (we have over 600+). 

So the issue boild down to this:
LSM, when rendered by RTL/Jest env=jsdom, returned an empty object on initialization since its running in nodejs first then loaded into jsdom. 

#### Solution: 

- overload the rtl render method [more deets here:](https://testing-library.com/docs/react-testing-library/setup)
- initialize the store in a wrapper that is rendered by the StateMachineProvider to ensure the lib is loaded by the time we init the Context  
- allow comp to receive props the same way it would be used in the app

**The setup:**
```
// test-utils.js 

import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore } from 'little-state-machine';
import Providers from 'components/Providers';

// wrapper should pass down props to 
const Wrapper = (props) => {
  const { component: Component } = props;
  createStore({ storeName: { some: 'value' } });

  return <Component {...props} />;
};

// allow providers to receive custom props if necessary
const render = (ui, { languageCode, ...options }) => {
  const AllTheProviders = ({ children }) => <Providers languageCode={languageCode}>{children}</Providers>;

  return rtlRender(ui, { wrapper: AllTheProviders, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render, Wrapper }; 
```

**Usage:**
```
import { render, Wrapper, screen, act } from 'lib/test-utils';
import MyTestComp from 'components/MyTestComp';

describe('<MyTestComp />', () => {
  test('renders Loader while fetching', async () => {
     
     // [act is required for handling state updates caused by LSM](https://reactjs.org/docs/test-utils.html#act)
    await act(async () => {
      render(<Wrapper component={MyTestCompnent} handleClick={jest.fn()} />, { languageCode: 'en-ca' });
    });

    const loaderElement = screen.getByText('Loader');  
    expect(loaderElement).toBeInTheDocument();
  });  
})
```

In summary, yes all tests should be wrapped by their own providers (which was original intent anyway), but LSM needs to initialize the store within a wrapper component so that the lib can load by the time the test comp gets rendered by RTL. 

I hope this helps save some time for others since there's literally 0 examples out there.
